src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -269,6 +277,222 @@ private Optional<Pocket> findPocketForReload(ReloadBalanceRequest reloadRequest,
+        log.info("Is need to return amount to new pocket: {}", isNeedToReturnAmountToNewPocket);
+        if (isNeedToReturnAmountToNewPocket) {
+            return revertToNewPocket(balance, revertRequest, chargingOperationToRevert, changeEvent);
+        } else {
+            return revertToOldPockets(chargingOperationToRevert, changeEvent);
+        }
+    }
+
+    private BalanceChangeEventDto revertToNewPocket(Balance balance,
+                                                    RevertBalanceOperationRequest revertRequest,
+                                                    BalanceChangeEventDto chargingOperationToRevert,
+                                                    BalanceChangeEvent changeEvent) {
+        Pocket pocket = new Pocket()
+            .key(randomUUID().toString())
+            .balance(balance)
+            .amount(chargingOperationToRevert.getAmountDelta())
+            .endDateTime(revertRequest.getEndDateTime())
+            .startDateTime(revertRequest.getStartDateTime())
+            .label(revertRequest.getLabel())
+            .metadata(Metadata.of(revertRequest.getMetadata()));
+
+        Pocket savedPocket = pocketRepository.save(pocket);
+        PocketChangeEvent event = createPocketChangeEvent(savedPocket, chargingOperationToRevert.getAmountDelta(), ZERO);
+
+        changeEvent.addPocketChangeEvent(event);
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private BalanceChangeEventDto revertToOldPockets(BalanceChangeEventDto chargingOperationToRevert,
+                                                     BalanceChangeEvent changeEvent) {
+        Map<Long, Pocket> pocketIdToPocketMap = new HashMap<>();
+
+        for (PocketChangeEventDto pocketChangeEventDto : chargingOperationToRevert.getPocketChangeEvents()) {
+            Pocket pocket = getPocketForUpdate(pocketChangeEventDto.getPocketId());
+            pocketIdToPocketMap.put(pocket.getId(), pocket);
+        }
+        for (PocketChangeEventDto pocketChangeEventDto : chargingOperationToRevert.getPocketChangeEvents()) {
+            Pocket pocket = pocketIdToPocketMap.get(pocketChangeEventDto.getPocketId());
+
+            BigDecimal amountBefore = pocket.getAmount();
+            BigDecimal amountToProvide = pocketChangeEventDto.getAmountDelta();
+
+            pocket.addAmount(amountToProvide);
+
+            Pocket savedPocket = pocketRepository.save(pocket);
+            PocketChangeEvent event = createPocketChangeEvent(savedPocket, amountToProvide, amountBefore);
+            changeEvent.addPocketChangeEvent(event);
+        }
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private void assertRevertOperationAlreadyPerformed(String uuidToRevert) {
```

Comment:
O_o maybe
assertRevertOperationNotPerformed

Reply:
??
Supposed that method assertRevertOperationAlreadyPerformed throws exception, when balance change event exist by uuid to revert.

Reply:
exception throws when action already performed

that`s mean we assert that operation not performed

Reply:
Okay, fixed